### ISSUE TITLE
differences in sun.print_params and SSW IDL get_sun results

### ISSUE THREAD

**Issue by jslavin** [sunpy/sunpy#1402]
Hi,

I've been looking at the results for solar parameters from the sun module and comparing them with those gotten from the SSW IDL routine get_sun.  The results are mostly close, though the heliographic longitude and latitude and (especially) the position angle of north pole are pretty far off.  Here's an example:

sun.print_params(date)
Solar Ephemeris for Sun Mar  1 00:16:36 2015

Distance (AU) = 0.990443193362
Semidiameter (arc sec) = 968.216053494 arcsec
True (long,lat) in degrees = (339.087367636,0.0)
Apparent (long, lat) in degrees = (339.082661918,0)
True (RA, Dec) = (-0.327494596065,0.934125799329
Apparent (RA, Dec) = (22.7117194929,-8.36845723654
Heliographic long. and lat of disk center in deg = ([353.81314120499883, -7.1973593104170925])
Position angle of north pole in deg = 40.8588966873
Carrington Rotation Number = 2161.0316597

IDL> results = get_sun(date,/list)
 Solar Ephemeris for  1-MAR-15  00:16:36
 
 Distance (AU) = 0.99068251
 Semidiameter (arc sec) = 968.65544
 True (long, lat) in degrees = (340.09175, 0.00000)
 Apparent (long, lat) in degrees = (340.08704, 0.00000)
 True (RA, Dec) in hrs, deg = (22.774631, -7.7839567)
 Apparent (RA, Dec) in hrs, deg = (22.774316, -7.7849442)
 Heliographic long. and lat. of disk center in deg = (354.82537, -7.2115739)
 Position angle of north pole in deg = -21.428021
 Carrington Rotation Number = 2161.0317

I'll look into the code a bit more, but maybe someone else can give me insight into these differences (and which is more accurate).

Jon

**Comment by Cadair** [comment]
ping @dpshelio @wafels @ehsteve

**Comment by dpshelio** [comment]
Thanks @jslavin I'm working in a new implementation of the ephemerides while documenting it and testing it on the way: #1070 
It's still work in progress but we will update this issue once we get it all done properly.

**Comment by jslavin** [comment]
I did find one typo in the code that affects the position angle of the
north pole:
224c224
<     x =
np.degrees(np.arctan(-np.cos(np.radians(lamda2)*np.tan(np.radians(ob1)))))
---
>     x =
np.degrees(np.arctan(-np.cos(np.radians(lamda2))*np.tan(np.radians(ob1))))

(misplaced parentheses).  This is for version 0.5.4.

Jon

**Comment by jslavin** [comment]
Hi,

I was just doing a little more checking of the results of limb placement on
XRT images.  I hadn't used the ssw map routines before, but I have done
that now for comparison.  With the particular image that I used, the ssw
routine places the limb correctly, but the sunpy map routine doesn't (see
attached images). It doesn't seem to help to provide the metadata, as I
have now done in the image fits file.  If you're interested I can supply
the code that used to produce the images as well as the data file. (Note: I
also tried using the rotate method on the map created by sunpy, but again
it doesn't seem to change anything.)

Regards,
Jon

**Comment by wafels** [comment]
Jonathan,

I don't think the images were attached.  Could you post the images, and the
SSWIDL/sunpy code snippets you're using?  That might help diagnose where
the issues are.  I took a look yesterday at the code in question and
nothing immediate jumped out at me.  There are online services where some
of the parameters of sun.py are calculated that we could use to check at
least those parameters.  That would help narrow down where the bugs are.
I'll take a closer look in the next few days.

Thanks,

Jack

**Comment by jslavin** [comment]
Hmm.  I had thought that I had posted them.  Here they are.

Jon
![sunpy_map](https://cloud.githubusercontent.com/assets/644963/7453472/8ee95ab8-f237-11e4-8435-92db4e5f03f6.png)
![idl_map](https://cloud.githubusercontent.com/assets/644963/7453480/a0f75a3e-f237-11e4-887c-3fa95cc7d6a8.png)

**Comment by jslavin** [comment]
And here are the code snippets:
SSWIDL:
file = 'L1_XRT20090711_000000.4.fits'
fits2map,file,map,header=header
loadct,3
plot_map,map,/limb,grid=15,drange=[-1.89,800.]
write_png,'idl_map.png',tvrd(/true)
end

sunpy:
import matplotlib.pyplot as plt
from sunpy import map
file = 'L1_XRT20090711_000000.4.fits'
m = map.Map(file)
m.rotate()  # this doesn't seem to make a difference
m.plot()
m.draw_limb()
m.draw_grid()
plt.savefig('sunpy_map.png')

Jon

**Comment by jslavin** [comment]
I tried to attach the fits file with the image, but that didn't work. Here's a link to the file:
https://www.dropbox.com/s/l932bzzxx9sbog2/L1_XRT20090711_000000.4.fits?dl=0

Jon

**Comment by Cadair** [comment]
It looks like the centre is wrong.

**Comment by jslavin** [comment]
I agree.  Using SSW IDL (get_sun) with the date for this image I get a heliographic latitude of 3.9356 vs. 3.8339 with sunpy.sun.  On the other hand, I created a fits header keyword HGLT_OBS and set it equal to the value from get_sun and this did not fix the problem.  The way sunpy sets the center seems very straightforward, but I'm not sure how ssw map does it.

**Comment by jslavin** [comment]
I think the issue is related to there being different values for XCEN, YCEN (used by SSWIDL routines) and CRVAL1, CRVAL2 (used by sunpy).  I'm not clear if this is an error in sunpy or maybe an error in the XRT fits headers.  Any information that you all can provide on this would be appreciated.

**Comment by jslavin** [comment]
This appears to actually be a problem with the XRT header keyword values for CRVAL1 and CRVAL2.  After a recent upgrade to the alignment in the pipeline, those values were not updated in the generated headers. The limb is now showing correctly.  

I think there are still issues with some of the solar parameters differing between solarsoft and sunpy, though now I'm not sure which are more accurate.

**Comment by Laubeee** [comment]
I am still getting slightly different values from `coordinates.sun.angular_radius` vs IDL `get_sun`
e.g. for 2014-02-07 I get 972.62224989 from sunpy vs 973.04023398772733 from IDL

**Comment by Laubeee** [comment]
if this changed on SSW side, maybe someone confirm/reproduce first, my SSW installation might not be up-to-date

**Comment by dstansby** [comment]
Can you let us know what version of sunpy you're using, and ideally a small self contained example that calculates those numbers for both sunpy and SSW so it's easier to track down any potential discrepancies?

**Comment by ayshih** [comment]
@Laubeee: ~95% of the discrepancy is due to the fact that SSW's `get_sun.pro` (indirectly) uses a photospheric radius of 696,000 m, and SunPy uses a photospheric radius of 695,700 m, which is the current IAU value.  There's nothing necessarily wrong if you want to work with a different value for the photospheric radius – e.g., AIA chooses to use 696,000 m – but you need to make sure all other calculations are consistent.  Where relevant, our coordinate frames include the frame attribute `rsun` to account for definitional differences.

~5% of the discrepancy is due to the fact that `get_sun.pro` is using less accurate formulae.  `get_sun.pro` implements Meeus's approximations that are intended for calculator use, and the resulting inaccuracy doesn't make a whole lot of sense on modern computers.  Accordingly, which SunPy agrees with the Astronomical Almanac on the Sun-Earth distance to the order of kilometers, `get_sun.pro` can be off by thousands of kilometers.

**Comment by Laubeee** [comment]
Thanks for clarifying :) this is good to know, i'll keep it in mind to use a consistent value.
